### ISSUE TITLE
Null reference exception during activation collection

### ISSUE THREAD

**Issue by josiahdecker** [dotnet/orleans#3165]
We're seeing an issue where activation collection on a silo seems to either stop or to take  a very long time.  The total activation count begins to increase at a steady rate while the recently used activation count stays steady.  Eventually this leads to the server running out of memory and crashing.  In the histogram below you can see logs for the "Before collection", "Starting DestroyActivations", and "After collection" events.  Typically they occur in quick successesion, but it seams that something happens that suddenly makes the process take a very long time.  Using the IManagementGrain interface to trigger a manual activation does seem to go quickly (but the timespan used in that case may not match the default lifetime of the grain).  We've only seen this in our production environment, we haven't been able to reproduce it in a test scenario.  It does seem to happen more frequently in high traffic scenarios.  

When the collections start to take a long time we also start to see an exception happen about once a minute:
Caught and ignored exception: System.NullReferenceException with mesagge: Object reference not set to an instance of an object. thrown from timer callback GrainTimer.Catalog.GCTimer. TimerCallbackHandler:[SystemTarget: S<ip-removed>:235355348Catalog@S0000000e]->System.Threading.Tasks.Task OnTimer(System.Object)

 at Orleans.Runtime.ActivationCollector.Bucket.CancelAll()
   at Orleans.Runtime.ActivationCollector.DequeueQuantum(IEnumerable`1& items, DateTime now)
   at Orleans.Runtime.ActivationCollector.ScanStale()
   at Orleans.Runtime.Catalog.<CollectActivationsImpl>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.GrainTimer.<ForwardToAsyncCallback>d__20.MoveNext()

It looks like the "CancellAll" call in Bucket doesn't check to see if the ActivationData object is null.  I assume that there is a concurrent call to Bucket.TryRemove, which is nulling out the ActivationData object in the dictionary.   

Two notable things about our setup.  First, we do have a very short lifetime on some of our grains.  The grain state can be large, so we want to remove it from memory quickly.  Originally we had it set to 15 minutes, then five, and eventually one minute.  I'm not sure if the above problem occurs more frequently with lower lifetimes.  Second, our grains are currenly sending very large messages in many cases, sometimes 1 mb or more.  We also get warning messages that activation turns are taking too long, though the warning is usually between 500 and 1000 milliseconds.

A process dump  of one of the servers in a bad state showed two worker threads actively processing messages, and the rest waiting on BlockingCollection<>s (we don't use those in our code so I assume that is the AsynchAgent queue).  The CPU averages around 50%.

![image](https://user-images.githubusercontent.com/120828/27487767-3f0d1720-57fa-11e7-8650-f008dcf0acc0.png)

**Comment by gabikliot** [comment]
Hmmmm... It's first time in maybe 2 years since I saw any problem with activation collector. I wonder if something changed recently or if your setup is really unusual.
I will try to look deeper.

What is your activation gc settings? Can you copy paste from the silo log when it prints the config at startup?
Can you share your full silo log?

**Comment by gabikliot** [comment]
OK, so there is definitely a concurrency bug in the bucket like you described - TryCancel (and not TryRemove) is nulling the value.

The fix will be trivial: there is no need in TryCancel at all. Just replace the one place where it is called in ScanAll with TryRemove.

This may fix your OOM problem. Lets try fixing it and see if the problem goes away. If not, we will need the logs to troubleshoot that.

**Comment by gabikliot** [comment]
Thought more about it: if the periodic scan to be too frequent (you set min activation GC to very low) and you have a lot of activation whose deactivation takes long time, it maybe that the next Scan starts before the prev finished (the Scan actually deactivates the grain, so if you are dojng something heavy in OnDeactivateAsyns it will slow the whole Scan). 

With your full logs I will be able to validate this. If this is indeed the case, we can pretty easy fix it, but conditioning the next scan not to start if there is on in progress. This will be much better than dealing with 2 concurrent scans.

**Comment by gabikliot** [comment]
Checked more: 2 scans are not supposed to ever run in parallel. This is already provided by the timer: the timer guarantees that the next tick does not occur before the prev finished. So what I described in the prev comment is not possible. In the log you should always see Before matched with After and never interleave.

The only way the above is not true is if you force the collection manually via the MgmntGrain. Which is what you do I guess.

**Comment by josiahdecker** [comment]
I don't have a deep familiarity with the code so please correct me if I have something wrong here.  In the case where a target grain already exists it looks like IncomingMessageAgent.ReceiveMessage will use that target grain's activation context to enqueue an incoming message (line 139, then down in EnqueueReceiveMessage on line 176).  This means that Dispatcher.ReceiveMessage will be on the target's context, not the system context.  That will call into ActivationCollector.TryRescheduleCollection which eventually makes it's way to Bucket.TryRemove which calls Bucket.TryCancel.

If that is correct then I assume that the TryCancel call which null's the activation data can execute concurrently with the grain collection code, since that grain collection code is on the system context.  It does seem unlikely since the grain context would need to get a reference to the bucket before it get's removed by the system context, then the system context would need to catch the grain after TryCancel has been called but before TryRemove has finished.  At the same time, if it is possible then it's just a statistics question that depends on how often the grains are collected and how often a message comes in for a grain that's in the pending collection bucket.  However, it does seem extremely unlikely that this proposed race condition explains the situation we saw where pretty much every collection was getting interrupted.

We did have a very low min GC (1 minute), which made the problem more likely.  And we were using the IManagementGrain at one point.  We can cut out the IManagementGrain usage, and we're raised the min grain lifetime to 15 minutes.  I don't know if that's still considered to short. 

You suggested replacing the call to TryCancel with TryRemove.  It looks like TryRemove actually calls TryCancel.  Would it also be safe to remove the code in TryCancel that nulls out the item in the dictionary?  There's a comment that notes it's just a perf optimization, and if it's called in the context of TryRemove then it seems superfluous. But the call to ConcurrentDictionary.TryUpdate does a reference comparison on the ActivationData object and I'm not sure what the significance of that is.

  I can monitor our logs to see if the situation has changed since we cut out the IManagementGrain.  If you have time to look at the raw logs I can send them to you.  I don't know if you would want the whole thing or just a specific subset.  Where can I submit them?

**Comment by jdom** [comment]
@benjaminpetit can you validate if you see similar side effects in the stuck grain prod issue that you were tracking down?

BTW, @gabikliot there is indeed something going on with collection (at least in Orleans 1.3.1), but I don't know at this point if this issue is related. We've been noticing very infrequently that some grain activations get stuck in the `Deactivating` state forever, so all messages to that grain end up timing out (and the NonReentrancyQueueSize growing forever) until the silo is restarted.

This is an example of the details for the stuck grain that @benjaminpetit gathered with OrleansManager:

```
**DetailedGrainReport for grain *grn/E241C918/00000000000000000009000006c5617803ffffffe241c918-0x65C2D945 from silo XXX.SiloWorker_IN_31 SiloAddress=S100.117.4.16:20002:233238197
   LocalCacheActivationAddresses=[]
   LocalDirectoryActivationAddresses=[]
   PrimaryForGrain=S100.117.2.29:20002:233242118
   GrainClassTypeName=XXX.FooGrain
   LocalActivations:
[      [Activation: S100.117.4.16:20002:233238197*grn/E241C918/00000000000000000009000006c5617803ffffffe241c918-0x65C2D945@51eeefb2 #GrainType=XXX.FooGrain Placement=RandomPlacement State=Deactivating NonReentrancyQueueSize=294 EnqueuedOnDispatcher=0 InFlightCount=0 NumRunning=0 IdlenessTimeSpan=7.01:47:56.2460274 CollectionAgeLimit=01:15:00]].
```

Notice that the IdlenessTimeSpan is of several days, and the queue size very large. This FooGrain does nothing in its `OnDeactivateAsync` method.

We also saw some of these in the logs, although it's hard to know if they are related, but seems to break some invariant in the activator collector logic:

```
ActivationCollector found an active activation in it's last bucket. This is violation of ActivationCollector invariants. For now going to defer it's collection. Activation: [Activation: S100.117.4.16:20002:233238197*grn/716E8E94/4b278ea34b0ad46ff8f2097bd0663b9706000000716e8e94+SMSProvider_XXX.ShoulderTap-0xFB5702BF@0cda6a91 #GrainType=Orleans.Streams.PubSubRendezvousGrain Placement=RandomPlacement State=Valid NonReentrancyQueueSize=0 EnqueuedOnDispatcher=0 InFlightCount=1 NumRunning=1 IdlenessTimeSpan=01:15:23.9608313 CollectionAgeLimit=01:15:00]
```

Notice that this is in the pub sub grain, not the FooGrain that gets stuck, but that FooGrain is using SMS. 

The code that produces that log entry is in [ActivationCollector#L235](https://github.com/dotnet/orleans/blob/56dd7842db5436c14e8d046f672bcf75e1c4ace8/src/OrleansRuntime/Catalog/ActivationCollector.cs#L235), and notice the code comment saying: `This is essentialy a bug, an active activation should not be in the last bucket.`

**Comment by gabikliot** [comment]
1) The race that causes null I think happens due to you calling mgmt grain. Anyway, it does not really matter. The race can be easily fixed.

2) I suggest we first fix the race: by removing TryCancel completely and inlining its logic in the Try Remove. No need to replace the pointer with null, just remove the activation completely from the dictionary. @jdom , @sergeybykov , can you guys do that?

3) After that is check in, @josiahdecker , you can try running again. No need to use mgmt grain again. you can configure the activation GC as you like. if the problem occurs again, I will need the full logs to investigate.

4) @jdom, the error from the collector is consistent with being stuck in de-activation mode. The condition says that the activation is !inactive (that is, it IS ACTIVE) but it is in the last bucket. Active activation  should always be in the 1st bucket (last bucket is the least recently used activations). I would imagine this (!activation.IsInactive) happens due to the activation executing the OnDeactivate and is thus "ACTIVE". 
Therefore, at least based on that one error it does not seems to be the issue with Activation GC but rather with why activation was stuck in OnDeactivate for ever.
Maybe some bug was introduced that causes this stucking? I do remember @sergeybykov opening this issue about stuck activation and @benjaminpetit issuing some warning/detection. It did sound strange to me, as I don't remember seeing similar things in the past.

**Comment by benjaminpetit** [comment]
No I don't think I see similar side effets in the stuck activation issue.

@josiahdecker do you know if the activation not collected are still accessible (ie you can make a call to them)?

**Comment by josiahdecker** [comment]
I'm not sure if they are still accessible.  We don't currently have verbose logging turned on so I haven't been able to identify the activations that are getting stuck.  Other than the trace messages is there some way to see which activations are currently being collected?

**Comment by sergeybykov** [comment]
Resolved via #3399.
